controllers/persistentvolumeclaim_controller.go

```diff
@@ -0,0 +1,331 @@
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// delete name from annotation.
+		_, nameFound := annotations[rsCronJobNameAnnotation]
+		if nameFound {
+			// remove name annotation by patching it to null.
+			patch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{%q: null}}}`, rsCronJobNameAnnotation))
+			err = r.Client.Patch(ctx, pvc, client.RawPatch(types.StrategicMergePatchType, patch))
+			if err != nil {
+				logger.Error(err, "Failed to remove annotation")
+
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("Annotation not set, exiting reconcile")
+		// no schedule annotation set, just dequeue.
+		return ctrl.Result{}, nil
+	}
+	logger = logger.WithValues("Schedule", schedule)
+
+	if rsCronJob != nil {
+		newRSCronJob := constructRSCronJob(rsCronJob.Name, req.Namespace, schedule, pvc.Name)
+		if reflect.DeepEqual(newRSCronJob.Spec, rsCronJob.Spec) {
+			logger.Info("No change in reclaimSpaceCronJob.Spec, exiting reconcile")
+
+			return ctrl.Result{}, nil
+		}
+		// update rsCronJob spec
+		rsCronJob.Spec = newRSCronJob.Spec
+		err = r.Client.Update(ctx, rsCronJob)
+		if err != nil {
+			logger.Error(err, "Failed to update reclaimSpaceCronJob")
+
+			return ctrl.Result{}, err
+		}
+		logger.Info("Successfully updated reclaimSpaceCronJob")
+
+		return ctrl.Result{}, nil
+	}
+
+	rsCronJobName := generateCronJobName(req.Name)
+	logger = logger.WithValues("ReclaimSpaceCronJobName", rsCronJobName)
+	// add cronjob name in annotations.
+	patch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{%q:%q}}}`, rsCronJobNameAnnotation, rsCronJobName))
+	err = r.Client.Patch(ctx, pvc, client.RawPatch(types.StrategicMergePatchType, patch))
+	if err != nil {
+		logger.Error(err, "Failed to update annotation")
+
+		return ctrl.Result{}, err
+	}
+
+	rsCronJob = constructRSCronJob(rsCronJobName, req.Namespace, schedule, pvc.Name)
+	err = ctrl.SetControllerReference(pvc, rsCronJob, r.Scheme)
+	if err != nil {
+		logger.Error(err, "Failed to set controllerReference")
+
+		return ctrl.Result{}, err
+	}
+
+	err = r.Client.Create(ctx, rsCronJob)
+	if err != nil {
+		logger.Error(err, "Failed to create reclaimSpaceCronJob")
+
+		return ctrl.Result{}, err
+	}
+	logger.Info("Successfully created reclaimSpaceCronJob")
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PersistentVolumeClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := mgr.GetFieldIndexer().IndexField(
+		context.Background(),
+		&csiaddonsv1alpha1.ReclaimSpaceCronJob{},
+		jobOwnerKey,
+		extractOwnerNameFromPVCObj)
+	if err != nil {
+		return err
+	}
+
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object == nil {
+				return false
+			}
+			// reconcile only if schedule annotation is found.
+			_, ok := e.Object.GetAnnotations()[rsCronJobScheduleTimeAnnotation]
+
+			return ok
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
```

Comment:
Does this not require the `update` permission for the PVC/annotations? `patch` is there, is it not cleaner to patch it in any case?

Reply:
Watch perm is enough to check for events.

Reply:
Ah, sure. Thanks!